File: backend/app/api/v1/evidence.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.dependencies import require_candidate
from app.api.errors import api_error
from app.db.session import get_db
from app.models.user import User
from app.schemas.evidence import EvidenceHubListResponse
from app.services.candidate import get_candidate_profile
from app.services.evidence_hub import (
    DEFAULT_LIMIT,
    MAX_LIMIT,
    EvidenceHubQuery,
    list_candidate_evidence,
)

router = APIRouter(prefix="/candidate", tags=["evidence"])


@router.get("/evidence", response_model=EvidenceHubListResponse)
def get_candidate_evidence(
    current_user: Annotated[User, Depends(require_candidate)],
    session: Annotated[Session, Depends(get_db)],
    source_type: Annotated[str | None, Query(max_length=64)] = None,
    skill: Annotated[str | None, Query(max_length=120)] = None,
    search: Annotated[str | None, Query(max_length=200)] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = DEFAULT_LIMIT,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> EvidenceHubListResponse:
    profile = get_candidate_profile(session, current_user.id)
    if profile is None:
        raise api_error(404, "PROFILE_NOT_FOUND", "Candidate profile not found")

    return list_candidate_evidence(
        session,
        candidate_id=profile.id,
        query=EvidenceHubQuery(
            source_type=source_type,
            skill=skill,
            search=search,
            limit=limit,
            offset=offset,
        ),
    )

File: frontend/lib/api/evidence.ts
import { apiRequest } from "@/lib/api/client";
import type {
  EvidenceHubListResponse,
  EvidenceHubQueryParams
} from "@/lib/api/types/evidence";

function toQueryString(params: EvidenceHubQueryParams): string {
  const search = new URLSearchParams();

  if (params.source_type) {
    search.set("source_type", params.source_type);
  }
  if (params.skill?.trim()) {
    search.set("skill", params.skill.trim());
  }
  if (params.search?.trim()) {
    search.set("search", params.search.trim());
  }
  if (params.limit != null) {
    search.set("limit", String(params.limit));
  }
  if (params.offset != null) {
    search.set("offset", String(params.offset));
  }

  const query = search.toString();
  return query ? `?${query}` : "";
}

export function listCandidateEvidence(
  params: EvidenceHubQueryParams = {}
): Promise<EvidenceHubListResponse> {
  return apiRequest<EvidenceHubListResponse>(`/candidate/evidence${toQueryString(params)}`);
}
